Fix units not being pulled from source sensor

rsteckler opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The current Reimann integration doesn't pull unit_of_measurement from the source sensor, as intended.  This is a regression from the previous functionality due to change https://github.com/home-assistant/core/pull/54802 .  While well intentioned, the change created a situation where two things happened:
1) The Reimann sensor would attempt to update it's own unit_of_measurement based on the source sensor, which wasn't ready yet.  This resulted in the Reimann sensor setting its units to a value that included the prefix and time, but not the source units.  Since the value of the Reimann units was no longer None, these never got updated once the source sensor units were actually available.  The end result of this is that Reimann sensors all have unit_of_measurement that were "h" instead of "Wh" in the case of a source sensor that used Watts.  
2) The intended state wasn't actually changed in the Reimann sensor because the function returned before the state was written.  This has been addressed using the local update_state variable.  

The reason this regression likely wasn't found earlier is because it only affects new sensors.  Anyone who set one up previously would have an existing unit_of_measurement from the database or the restore_state.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://community.home-assistant.io/t/reimann-integration-has-wrong-unit-of-measurement/376402
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rsteckler committed:
Fixed units not being pulled from source sensor

homeassistant commented:
Hi @rsteckler, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#63639).

Once you do that we will be able to review and accept this pull request.

Thanks!

dgomes commented on homeassistant/components/integration/sensor.py:
```diff
@@ -150,17 +150,29 @@ def calc_integration(event):
-                )
+                if unit is not None:
+                    self._unit_of_measurement = self._unit_template.format(
+                        "" if unit is None else unit
```
This does not make sense... you are inside a block where `unit is not None`

rsteckler replied:
Good point.  I was just wrapped the existing assignment into a block to protect it from unit == None.  Now that it's protected, we can assume it's not None during the assignment.  
I'll remove the redundant check.

rsteckler replied:
The problem with the original line, just to be clear, is that when unit was None, there was still an assignment to self._unit_of_measurement, placing a "" if unit was None.  Therefore self._unit_of_measurement was not None anymore and this assignment didn't execute on the next pass (when unit IS populated).

rsteckler replied:
This has been updated.

dgomes commented on homeassistant/components/integration/sensor.py:
```diff
@@ -150,17 +150,29 @@ def calc_integration(event):
+                update_state = True
+
+            if update_state:
+                self.async_write_ha_state()
```
Why not just call `self.async_write_ha_state()` ?

rsteckler replied:
I'm new to HA and don't have a good sense for how expensive async_write_ha_state() is.  The update_state = true assignment can happen in two cases above and I want to call update if one or both cases happened.  If it's very cheap to call async_write_ha_state(), I can just call it in both places above.  Would appreciate your guidance on that.

dgomes replied:
probably you can call `self.async_write_ha_state()` regardless at the end

rsteckler replied:
There's a conditional and a potential return statement that leaves the function right below this.  I wanted to make sure it updates state before it leaves the function.
The check and return always existed, as far as I can tell.  The reason these two checks (units and device class) are above the return is because the previous author wanted to make sure the setting of units and device class happened even if the source sensor hadn't updated yet.  They just neglected to call async_wite_ha_state(), so the desired updates never actually worked.  
I see three options:
1) Leave it broken or handle that issue in a separate PR
2) Call async_write_ha_state multiple times (up to three) on execution of this function
3) Use a temp local to track whether it needs to be called (once) before the return
4) Wrap the entire method in a giant conditional so it only executes when the check for a valid state passes.  This is conceptually the same as the check/return, but introduces a level of indentation that I think is less clean.

I prefer options 2 or 3, but again, this is my first PR to HA, so very much open to your feedback.

dgomes replied:
OK lets keep the local temp

dgomes commented:
Have you tested this ?

rsteckler commented:
> Have you tested this ?

yes, locally with both existing (pre-change) sensors, as well as new sensors created after the change.
This doesn't retro-actively fix old Reimann sensors that have statistics that already have a bad unit (e.g. "h"), but it makes sure all new Reimann sensors work correctly.

rsteckler committed:
Removed redundant check for unit being None

dgomes commented:
> > Have you tested this ?
> 
> yes, locally with both existing (pre-change) sensors, as well as new sensors created after the change. This doesn't retro-actively fix old Reimann sensors that have statistics that already have a bad unit (e.g. "h"), but it makes sure all new Reimann sensors work correctly.

Please add a test

rsteckler committed:
Added test to verify Reimann sensor gets correct units

rsteckler commented:
> > > Have you tested this ?
> > 
> > 
> > yes, locally with both existing (pre-change) sensors, as well as new sensors created after the change. This doesn't retro-actively fix old Reimann sensors that have statistics that already have a bad unit (e.g. "h"), but it makes sure all new Reimann sensors work correctly.
> 
> Please add a test

I've added a test

> > > Have you tested this ?
> > 
> > 
> > yes, locally with both existing (pre-change) sensors, as well as new sensors created after the change. This doesn't retro-actively fix old Reimann sensors that have statistics that already have a bad unit (e.g. "h"), but it makes sure all new Reimann sensors work correctly.
> 
> Please add a test

Added a unit test that fails with the original code and passes with the new code.
I've learned that the functionality to pass through the correct units was actually being tested, but because the actual runtime sequence is different than the unit test setup sequence, the bug wasn't happening.
To be more specific, when HA starts up, it sends a series of updates to the Reimann sensor of the source sensor "being prepared".  This includes a transition of state from None to Unknown, then from Unknown to the attributes being populated, then later some actual state value changes.

The unit test framework doesn't replicate this implementation detail. Instead, it just passes the completely prepared sensor to the Reimann sensor on the first pass, making the defensive code in this PR work during unit test but fail during actual runtime.

The new unit test replicates the runtime behavior by intentionally sending sensor updates when the source sensor has state, but no attributes yet.  This replicates the runtime issue.  I don't love the idea of replicating what's almost certainly an implementation detail in the HA startup sequence, but it effectively exercises the defensive code and if the startup sequence changes to something else in the future and this test (and the defensive code it tests) become obsolete, they at least won't hurt anything.

dgomes reviewed (approved):
Coverage is bellow threshold

rsteckler committed:
Added tests to exercise legacy calc errors and device class decision

rsteckler commented:
> Coverage is bellow threshold

Added some tests to cover the new code, as well as some of the legacy code (exceptions)

rsteckler commented:
Note sure who to ask for help on this, but codecov/patch is showing that the new line 171 in sensor.py isn't being covered.  That line is a blank line, which is weird.  Also, lines 170 and 172 definitely are being covered by tests.

dgomes commented:
Codecov sometimes has some hiccups... let's wait for a 2nd reviewer and we move forward

frenck reviewed (approved):
Thanks, @rsteckler 👍
